src/kafka-net/KafkaTcpSocket.cs

```diff
@@ -337,8 +364,14 @@ private void ProcessNetworkstreamTasks(NetworkStream netStream)
                 {
                     if (OnReconnectionAttempt != null) OnReconnectionAttempt(attempts++);
                     _client = new TcpClient();
-                    await _client.ConnectAsync(_endpoint.Endpoint.Address, _endpoint.Endpoint.Port).ConfigureAwait(false);
+
+                    var connectTask = _client.ConnectAsync(_endpoint.Endpoint.Address, _endpoint.Endpoint.Port);
+                    await Task.WhenAny(connectTask, _disposeTask).ConfigureAwait(false);
+                    if (_disposeToken.IsCancellationRequested)
+                        throw new ObjectDisposedException(" on ReEstablishConnectionAsync Object is disposing.");
+
                     _log.WarnFormat("Connection established to:{0}.", _endpoint);
+                    await connectTask;//for ex to throw!!
```

Comment:
The MSDN documentation for WhenAny states:

The returned task will complete when any of the supplied tasks has completed. The returned task will always end in the RanToCompletion state with its Result set to the first task to complete. This is true even if the first task to complete ended in the Canceled or Faulted state.
Essentially what is happening is that the task returned by WhenAny simply swallows the faulted task. It only cares about the fact that the task is finished, not that it has successfully completed. When you await the task, it simply completes without error, because it is the internal task that has faulted, and not the one you're awaiting.